pilot/pkg/networking/core/v1alpha3/listener_waypoint.go

```diff
@@ -58,16 +58,50 @@ func (lb *ListenerBuilder) serviceForHostname(name host.Name) *model.Service {
 
 func (lb *ListenerBuilder) buildWaypointInbound() []*listener.Listener {
 	listeners := []*listener.Listener{}
-	// We create 3 listeners:
-	// 1. Decapsulation CONNECT listener.
-	// 2. IP dispatch listener, handling both VIPs and direct pod IPs.
-	// 3. Encapsulation CONNECT listener, originating the tunnel
-	wls, wps := findWaypointResources(lb.node, lb.push)
 
-	listeners = append(listeners,
-		lb.buildWaypointInboundConnectTerminate(),
-		lb.buildWaypointInternal(wls, wps.orderedServices),
-		buildWaypointConnectOriginateListener())
+	services := findWaypointServices(lb.node, lb.push)
+	workloads := findWaypointWorkloads(lb.node, lb.push)
+
+	// TODO use ServiceInfo and WorkloadInfo instead of using GetListener
+	waypoint := findWaypointInfo(lb.node, lb.push)
+
+	catchall := waypoint.GetBinding("*")
+	if catchall != nil && catchall.Protocol == protocol.HBONE {
+		// This is a native HBONE gatway, all services/workloads are handled in one internal listener.
+		// We create 3 listeners:
+		// 1. Decapsulation CONNECT listener.
+		// 2. IP dispatch listener, handling both VIPs and direct pod IPs.
+		// 3. Encapsulation CONNECT listener, originating the tunnel
+		return []*listener.Listener{
+			lb.buildWaypointInboundConnectTerminate(),
+			lb.buildWaypointInternal(workloads, services.orderedServices),
+			buildWaypointConnectOriginateListener(),
+		}
+	}
+
+	servicesByListener := map[model.WaypointBinding][]*model.Service{}
```

Comment:
After looking back at this, it might be excessively dynamic if we intend to eventually replace it with standard Gateway xDS anyway. 

Maybe we do something much simpler like  `PROXY on *` or `HBONE on *` triggering this stuff to get generated and the other listeners with hostname matchers are just doing filtering and trigger internal listeners.